src/structures/gdt.rs

```diff
@@ -184,55 +184,107 @@ pub enum Descriptor {
+        const LIMIT_16_19       = 0xF << 48;
+        /// Bits `0..=23` of the base field (ignored in 64-bit mode)
+        const BASE_0_23         = 0xFF_FFFF << 16;
+        /// Bits `24..=31` of the base field (ignored in 64-bit mode)
+        const BASE_24_31        = 0xFF << 56;
     }
 }
 
+/// The following constants define default values for common GDT use-cases. They
+/// are all "flat" segments meaning that they can access the entire address
+/// space. These values all set [`WRITABLE`][DescriptorFlags::WRITABLE] and
+/// [`ACCESSED`][DescriptorFlags::ACCESSED].
+///
+/// In short, these values disable segmentation, permission checks, and access
+/// tracking at the GDT level. Kernels using these values should use paging to
+/// implement this functionality.
+impl DescriptorFlags {
+    // Flags that we set for all user segments
+    const COMMON: Self = Self::from_bits_truncate(
+        Self::USER_SEGMENT.bits()
+            | Self::PRESENT.bits()
+            | Self::WRITABLE.bits()
+            | Self::ACCESSED.bits(),
+    );
+    // Flags that need to be set for all flat 32-bit segments
+    const COMMON_FLAT: Self = Self::from_bits_truncate(
+        Self::COMMON.bits()
+            | Self::LIMIT_0_15.bits()
+            | Self::LIMIT_16_19.bits()
+            | Self::GRANULARITY.bits(),
+    );
+    /// A kernel data segment (64-bit or flat 32-bit)
+    pub const KERNEL_DATA: Self =
+        Self::from_bits_truncate(Self::COMMON_FLAT.bits() | Self::DEFAULT_SIZE.bits());
+    /// A flat 32-bit kernel code segment
+    pub const KERNEL_CODE32: Self = Self::from_bits_truncate(
+        Self::COMMON_FLAT.bits() | Self::EXECUTABLE.bits() | Self::DEFAULT_SIZE.bits(),
+    );
+    /// A 64-bit kernel code segment
+    pub const KERNEL_CODE64: Self = Self::from_bits_truncate(
+        Self::COMMON.bits() | Self::EXECUTABLE.bits() | Self::LONG_MODE.bits(),
+    );
+    /// A user data segment (64-bit or flat 32-bit)
+    pub const USER_DATA: Self =
+        Self::from_bits_truncate(Self::KERNEL_DATA.bits() | Self::DPL_RING_3.bits());
+    /// A flat 32-bit user code segment
+    pub const USER_CODE32: Self =
+        Self::from_bits_truncate(Self::KERNEL_CODE32.bits() | Self::DPL_RING_3.bits());
+    /// A 64-bit user code segment
+    pub const USER_CODE64: Self =
+        Self::from_bits_truncate(Self::KERNEL_CODE64.bits() | Self::DPL_RING_3.bits());
+}
+
 impl Descriptor {
-    /// Creates a segment descriptor for a long mode kernel code segment.
+    /// Creates a segment descriptor for a 64-bit kernel code segment.
     #[inline]
-    pub fn kernel_code_segment() -> Descriptor {
-        use self::DescriptorFlags as Flags;
-
-        let flags = Flags::USER_SEGMENT | Flags::PRESENT | Flags::EXECUTABLE | Flags::LONG_MODE;
-        Descriptor::UserSegment(flags.bits())
+    pub const fn kernel_code_segment() -> Descriptor {
+        Descriptor::UserSegment(DescriptorFlags::KERNEL_CODE64.bits())
     }
```

Comment:
We should probably also add a `kernel_data_segment` function to resolve #160.

Also, a general `from_flags(flags: DescriptorFlags) -> Self` function be useful.

Reply:
Good catch I forgot about syscall/sysenter/sysret/sysexit, this might actually impact what we do for defaults here.

I'll do some more reading

Reply:
So it looks like we will should set all of these "ignored" flags. Per the Intel manual, syscall/sysenter/sysret/sysexit all load a specific selector from the GDT, but they don't actually read the GDT entry, they instead load fixed data in the descriptor caches. The manual then states:

>  It is the responsibility of OS software to ensure that the descriptors (in GDT or LDT) referenced by those selector values correspond to the fixed values loaded into the descriptor caches; the SYSCALL instruction does not ensure this
correspondence.

So we should just have our defaults match those fixed values the processor loads automatically. These are the values I had initially (i.e. exactly the values Linux uses). It now also makes sense why Linux was setting ignored bits, accessed bits, etc...

<details>
  <summary>Specific details of what flags are set by the processor</summary>

## Ring 3 -> Ring 0

### CS
  SYSCALL: SEL = IA32_START[47:32]
  SYSENTER: SEL = IA32_SYSENTER_CS[15:0]
  - Base = 0
  - Limit = 0xFFFFF
  - Type  = 0x1011 (execute, read, accessed)
  - S     = 1
  - DPL   = 0
  - P     = 1
  - L     = 1 (0 32-bit)
  - D     = 0 (1 32-bit)
  - G     = 1

### SS
  SYSCALL: SEL = IA32_START[47:32] + 8
  SYSENTER: SEL = IA32_SYSENTER_CS[15:0] + 8
  - Base = 0
  - Limit = 0xFFFFF
  - Type  = 0x0011 (data, write, accessed)
  - S     = 1
  - DPL   = 0
  - P     = 1
  - B     = 1
  - G     = 1

## Ring 0 -> Ring 3

### CS 
  SYSRET:          SEL = IA32_STAR[63:48] + 16
  SYSRET (32-bit): SEL = IA32_STAR[63:48]
  SYSEXIT:          SEL = IA32_SYSENTER_CS[15:0] + 32
  SYSEXIT (32-bit): SEL = IA32_SYSENTER_CS[15:0] + 16
  - Base = 0
  - Limit = 0xFFFFF
  - Type  = 0x1011 (execute, read, accessed)
  - S     = 1
  - DPL   = 3
  - P     = 1
  - L     = 1 (0 32-bit)
  - D     = 0 (1 32-bit)
  - G     = 1

### SS
  SYSRET:        SEL = IA32_STAR[63:48]+8
  SYSEXIT:          SEL = IA32_SYSENTER_CS[15:0] + 40
  SYSEXIT (32-bit): SEL = IA32_SYSENTER_CS[15:0] + 24
  - Base = 0
  - Limit = 0xFFFFF
  - Type  = 0x0011 (data, write, accessed)
  - S     = 1
  - DPL   = 3
  - P     = 1
  - B     = 1
  - G     = 1
</details>

Interestingly, due to the constraints of these commands, if you wanted to support both, you would need a GDT that had the following segments ordered like:

- KERNEL_CODE64
- KERNEL_DATA
- USER_CODE32
- USER_DATA
- USER_CODE64
- USER_DATA (only need for 64-bit sysenter/sysexit)

Reply:
> Also, a general `from_flags(flags: DescriptorFlags) -> Self` function be useful.

A user can already do `Descriptor::UserSegment(flags.bits())` do you think adding another method is useful? Should it do some sort of basic checks?

Reply:
> > Also, a general `from_flags(flags: DescriptorFlags) -> Self` function be useful.
> 
> A user can already do `Descriptor::UserSegment(flags.bits())` do you think adding another method is useful? Should it do some sort of basic checks?

I was mostly thinking about improving the discoverability of the `DescriptorFlags` structure. Right now, the relationship between the `DescriptorFlags` and `GlobalDescriptorTable` types is only implicit. You have to know the GDT details to be sure that `Descriptor::UserSegment(flags.bits())` is the correct implementation (as you could put any `u64` in there). An additional `from_flags` method would properly encode this at the type system level, which would make it less dangerous to use.

Reply:
> So we should just have our defaults match those fixed values the processor loads automatically. These are the values I had initially (i.e. exactly the values Linux uses). It now also makes sense why Linux was setting ignored bits, accessed bits, etc...

Thanks a lot for investigating this! I agree that it makes sense to use the sames values at Linux then.



> Interestingly, due to the constraints of these commands, if you wanted to support both, you would need a GDT that had the following segments ordered like:
> 
>     * KERNEL_CODE64
> 
>     * KERNEL_DATA
> 
>     * USER_CODE32
> 
>     * USER_DATA
> 
>     * USER_CODE64
> 
>     * USER_DATA (only need for 64-bit sysenter/sysexit)

Unrelated to this PR, but maybe it makes sense to add a new constructor function to `GlobalDescriptorTable` with exactly these entries? It seems like this is the canonical GDT layout that most users will want.

Reply:
> An additional `from_flags` method would properly encode this at the type system level, which would make it less dangerous to use.

Agreed here, but to make this `const_fn` we will have to add additional features. I'm going to merge this as-is and add any additional constructors in a follow-up PR (which will also add new GDT constructor).